### PR TITLE
MemoryLifetimeVerifier: treat enum cases with empty payloads (e.g. an empty tuple) like cases with no payloads.

### PR DESCRIPTION
**Pull request by eeckstein**
This PR is intended to replace https://github.com/apple/swift/pull/39173.

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci clean test linux

File: test/SIL/memory_lifetime_failures.sil [hunk 1]
```diff
@@ -567,3 +567,36 @@ bb1:
...
+  inject_enum_addr %0 : $*Result<Int, Error>, #Result.failure!enumelt
+  br bb3
+
+bb3:
+  %12 = load [take] %0 : $*Result<Int, Error>
+  destroy_value %12 : $Result<Int, Error>
+  dealloc_stack %0 : $*Result<Int, Error>
+  %15 = tuple ()
+  return %15 : $()
+}
```
Comment (meg-gupta) [inline test/SIL/memory_lifetime_failures.sil:602]: @eeckstein what about this slightly different SIL ?

```
sil [ossa] @test_init_enum_trivial_case : $@convention(thin) () -> @error Error {
bb0:
  %0 = alloc_stack $Result<Int, Error>
  %1 = function_ref @try_get_error : $@convention(thin) () -> @error Error
  try_apply %1() : $@convention(thin) () -> @error Error, normal bb1, error bb2

bb1(%3 : $()):
  %6 = init_enum_data_addr %0 : $*Result<Int, Error>, #Result.success!enumelt
  inject_enum_addr %0 : $*Result<Int, Error>, #Result.success!enumelt
  br bb3


bb2(%7 : @owned $Error):
  %8 = init_enum_data_addr %0 : $*Result<Int, Error>, #Result.failure!enumelt
  store %7 to [init] %8 : $*Error
  inject_enum_addr %0 : $*Result<Int, Error>, #Result.failure!enumelt
  br bb3

bb3:
  %12 = load [take] %0 : $*Result<Int, Error>
  destroy_value %12 : $Result<Int, Error>
  dealloc_stack %0 : $*Result<Int, Error>
  %15 = tuple ()
  return %15 : $()
}
```

Looks like we raise an error when we have no `init_enum_data_addr`. But bail out if we have `init_enum_data_addr` ?

Comment (meg-gupta) [reply test/SIL/memory_lifetime_failures.sil:602]: If this is out of scope for the PR, can you add a central comment to document unhandled cases and cases that need not be analyzed by the MemoryLifetimeVerifier and maybe a `memory_lifetime_unhandled.sil`   ? IDK if it exists already and I missed it?

Comment (eeckstein) [reply test/SIL/memory_lifetime_failures.sil:602]: Currently, the memory verifier can only handle single-payload enums.

Comment (eeckstein) [reply test/SIL/memory_lifetime_failures.sil:602]: In general, the MemoryLifetimeVerifier does only analyze a certain subset of locations, e.g. locations which are not aliased and locations which involve certain subset of all the address-instructions.
It's not really feasible to document which cases are _not_ supported. I added a new commit which improves the comment of `MemoryLocations` (and pointing to that in the comment of the MemoryLifetimeVerifier). This should hopefully make it a bit clearer.

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Comment by meg-gupta** [comment]
lgtm

File: test/SIL/memory_lifetime_failures.sil (hunk 1 above)
Comment (meg-gupta) [reply test/SIL/memory_lifetime_failures.sil:602]: Thanks

**Pull request merged**